doc/reference/serialization.rst

```diff
@@ -0,0 +1,267 @@
+   "74s", 20, "str (length 74)"
+   "c20s", 21, "[unsigned byte, str (length 20)]"
+   "bits", 1, "[bit 0, bit 1, bit 2, bit 3, bit 4, bit 5, bit 6, bit 7]"
+   "raw", "?", "str (length ?)"
+   "varlenBx2", "1 + ? * 2", "[str (length = 2), \.\.\. ] (length < 256)"
+   "varlenH", "2 + ?", "str (length ? < 65356)"
+   "varlenHx20", "2 + ? * 20", "[str (length = 20), \.\.\. ] (length < 65356)"
+   "varlenI", "4 + ?", "str (length < 4294967295)"
+   "doublevarlenH", "2 + ?", "str (length ? < 65356)"
+   "payload", "2 + ?", "Serializable"
+
+
+Some of these data types represent common usage of serializable classes:
+
+
+.. csv-table:: Common data types
+   :header: "member", "description"
+   :widths: 5, 20
+
+   "4SH", "(IP, port) tuples"
+   "20s", "SHA-1 hashes"
+   "32s", "LibNaCL signatures"
+   "64s", "LibNaCL public keys"
+   "74s", "LibNaCL public keys with prefix"
+
+
+Special instances are the ``raw`` and ``payload`` data types.
+
+- ``raw``: can only be used as the last element in a format list as it will consume the remainder of the input string (avoid if possible).
+- ``payload``: will nest another ``Serializable`` instance into this instance. When used, the ``format_list`` should specify the class of the nested ``Serializable`` and the ``to_pack_list()`` output should give a tuple of ``("payload", the_nested_instance)``. The ``VariablePayload`` automatically infers the ``to_pack_list()`` for you. See the ``NestedPayload`` class definition for more info.
+
+
+The ez_pack family for Community classes
+----------------------------------------
+
+All subclasses of the ``EZPackOverlay`` class (most commonly subclasses of the ``Community`` class) have a short-cut for serializing messages belonging to the particular overlay.
+This standardizes the prefix and message ids of overlays.
+Concretely, it uses the first 23 bytes of each packet to handle versioning and routing (demultiplexing) packets to the correct overlay.
+
+The ``ezr_pack`` method of ``EZPackOverlay`` subclasses takes an (integer) message number and a variable amount of ``Serializable`` instances.
+Optionally you can choose to not have the message signed (supply the ``sig=True`` or ``sig=False`` keyword argument for respectively a signature or no signature over the packet).
+
+The ``lazy_wrapper`` and ``lazy_wrapper_unsigned`` decorators can then respectively be used for unserializing payloads which are signed or not signed.
+Simply supply the payload classes you wish to unserialize to, to the decorator.
+
+As some internal messages and deprecated messages use some of the message range, you have the messages identifiers from 0 through 234 available for your custom message definitions.
+Once you register the message handler and have the appropriate decorator on the specified handler method your overlay can communicate with the Internet.
+In practice this will look something like this example:
+
+
+.. code-block:: python
+
+    class MyCommunity(Community):
```

Comment:
Maybe it would be good to have this example community `copy/paste` -able by providing a dummy MyMessagePayload1`, `MyMessagePayload2` and MASTER_KEY. I can imagine people trying to copy/paste the community code and run it. Also maybe put a link around somewhere to your earlier overlay tutorial. Just a thought.

Reply:
A runnable class would probably distract from the core message, I'll put in a link to the overlay tutorial for a copy/paste runnable example.